TypeError in test collection when sys.stdin isn't an iterator

In [this build](https://travis-ci.org/jaraco/jaraco.financial/jobs/354002885), I found that because of the construction of a `csv.DictReader(sys.stdin)` during import, the file can't be collected.

Looking a `--pdb` trace on it, I see:

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
jaraco/financial/lifo.py:82: in <module>
    input: compose(csv.DictReader, open)=csv.DictReader(sys.stdin),
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py:87: in __init__
    self.reader = reader(f, dialect, *args, **kwds)
E   TypeError: argument 1 must be an iterator
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py(87)__init__()
-> self.reader = reader(f, dialect, *args, **kwds)
(Pdb) f
<_pytest.capture.DontReadFromInput object at 0x1035952e8>
```

Obviously, I could refactor the code to bypass the error, but since the code runs, I feel it should be supported for pytest collection.

Perhaps `DontReadFromInput` could be adapted to be an iterator to satisfy csv.DictReader.

The issue can be minimally replicated with

```
$ cat > test.py
import sys
import csv
csv.DictReader(sys.stdin)
$ rwt -q pytest -- -m pytest test.py
======================================================== test session starts =========================================================
platform darwin -- Python 3.7.0b2, pytest-3.4.2, py-1.5.2, pluggy-0.6.0
rootdir: /Users/jaraco/clean, inifile:
collected 0 items / 1 errors

=============================================================== ERRORS ===============================================================
______________________________________________________ ERROR collecting test.py ______________________________________________________
test.py:3: in <module>
    csv.DictReader(sys.stdin)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py:87: in __init__
    self.reader = reader(f, dialect, *args, **kwds)
E   TypeError: argument 1 must be an iterator
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
====================================================== 1 error in 0.09 seconds =======================================================
```

I thought maybe adding `__next__` to DontReadFromInput would make it an iterator:

```
$ git diff
diff --git a/_pytest/capture.py b/_pytest/capture.py
index 36658acc..8d7ba6e7 100644
--- a/_pytest/capture.py
+++ b/_pytest/capture.py
@@ -566,6 +566,7 @@ class DontReadFromInput(object):
     readline = read
     readlines = read
     __iter__ = read
+    __next__ = read

     def fileno(self):
         raise UnsupportedOperation("redirected stdin is pseudofile, "
```

but it didn't:

```
jaraco/financial/lifo.py:82: in <module>
    input: compose(csv.DictReader, open)=csv.DictReader(sys.stdin),
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py:87: in __init__
    self.reader = reader(f, dialect, *args, **kwds)
E   TypeError: argument 1 must be an iterator
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/csv.py(87)__init__()
-> self.reader = reader(f, dialect, *args, **kwds)
(Pdb) f
<_pytest.capture.DontReadFromInput object at 0x11013e390>
(Pdb) f.__next__
<bound method DontReadFromInput.read of <_pytest.capture.DontReadFromInput object at 0x11013e390>>
```

--- a/_pytest/capture.py
+++ b/_pytest/capture.py
@@ -560,7 +560,7 @@ class SysCaptureBinary(SysCapture):
         return res
 
 
-class DontReadFromInput(object):
+class DontReadFromInput(six.Iterator):
     """Temporary stub class.  Ideally when stdin is accessed, the
     capturing should be turned off, with possibly all data captured
     so far sent to the screen.  This should be configurable, though,

--- a/_pytest/capture.py
+++ b/_pytest/capture.py
@@ -574,7 +574,10 @@ class DontReadFromInput(object):
         raise IOError("reading from stdin while output is captured")
     readline = read
     readlines = read
-    __iter__ = read
+    __next__ = read
+
+    def __iter__(self):
+        return self
 
     def fileno(self):
         raise UnsupportedOperation("redirected stdin is pseudofile, "

--- a/_pytest/logging.py
+++ b/_pytest/logging.py
@@ -292,8 +292,7 @@ def caplog(request):
     * caplog.text()          -> string containing formatted log output
     * caplog.records()       -> list of logging.LogRecord instances
     * caplog.record_tuples() -> list of (logger_name, level, message) tuples
-    * caplog.clear()         -> clear captured records and formatted log output
-                                string
+    * caplog.clear()         -> clear captured records and formatted log output string
     """
     result = LogCaptureFixture(request.node)
     yield result

--- a/changelog/3314.bugfix.rst
+++ b/changelog/3314.bugfix.rst
@@ -1,0 +1,3 @@
+During test collection, when stdin is not allowed to be read, the
+``DontReadFromStdin`` object still allow itself to be iterable and
+resolved to an iterator without crashing.

--- a/testing/test_capture.py
+++ b/testing/test_capture.py
@@ -751,7 +751,8 @@ def test_dontreadfrominput():
     assert not f.isatty()
     pytest.raises(IOError, f.read)
     pytest.raises(IOError, f.readlines)
-    pytest.raises(IOError, iter, f)
+    iter_f = iter(f)
+    pytest.raises(IOError, next, iter_f)
     pytest.raises(UnsupportedOperation, f.fileno)
     f.close()  # just for completeness
 
@@ -764,7 +765,8 @@ def test_dontreadfrominput_buffer_python3():
     assert not fb.isatty()
     pytest.raises(IOError, fb.read)
     pytest.raises(IOError, fb.readlines)
-    pytest.raises(IOError, iter, fb)
+    iter_f = iter(f)
+    pytest.raises(IOError, next, iter_f)
     pytest.raises(ValueError, fb.fileno)
     f.close()  # just for completeness
 

